src/main/host/syscall/poll.c

```diff
@@ -0,0 +1,241 @@
+    for (nfds_t i = 0; i < nfds; i++) {
+        struct pollfd* pfd = &fds[i];
+
+        if (pfd->fd < 0) {
+            continue;
+        }
+
+        CompatDescriptor* cdesc = process_getRegisteredCompatDescriptor(sys->process, pfd->fd);
+        utility_assert(cdesc); // we would have returned POLLNVAL in getPollEvents
+
+        struct epoll_event epev = {0};
+        if (pfd->events & POLLIN) {
+            epev.events |= EPOLLIN;
+        }
+        if (pfd->events & POLLOUT) {
+            epev.events |= EPOLLOUT;
+        }
+
+        if (epev.events) {
+            epoll_control(sys->epoll, EPOLL_CTL_ADD, pfd->fd, cdesc, &epev);
+        }
+    }
+}
+
+static SysCallReturn _syscallhandler_pollHelper(SysCallHandler* sys, PluginPtr fds_ptr, nfds_t nfds,
+                                                const struct timespec* timeout) {
+    // Get the pollfd struct in our memory
+    struct pollfd* fds =
+        process_getWriteablePtr(sys->process, sys->thread, fds_ptr, nfds * sizeof(*fds));
+
+    // Check if any of the fds have events now
+    int num_ready = _syscallhandler_getPollEvents(sys, fds, nfds);
+
+    debug("poll update: %i of %lu fds are ready", num_ready, nfds);
+
+    // Block or not depending on the timeout values
+    if (num_ready == 0) {
+        bool dont_block = timeout && timeout->tv_sec == 0 && timeout->tv_nsec == 0;
+
+        if (dont_block || _syscallhandler_wasBlocked(sys)) {
+            debug("No events are ready and poll needs to return now");
+            goto done;
+        } else {
+            debug("No events are ready and poll needs to block");
+
+            // Our epoll will tell us when we have events
+            _syscallhandler_registerPollFDs(sys, fds, nfds);
+
+            bool need_timer = timeout && (timeout->tv_sec > 0 || timeout->tv_nsec > 0);
+            if (need_timer) {
+                _syscallhandler_setListenTimeout(sys, timeout);
+            }
+
+            // Block on epoll, which is readable when any fds have events
+            Trigger trigger = (Trigger){.type = TRIGGER_DESCRIPTOR,
+                                        .object = (LegacyDescriptor*)sys->epoll,
+                                        .status = STATUS_DESCRIPTOR_READABLE};
+
+            // We either use our timer as a timeout, or no timeout
+            return (SysCallReturn){
+                .state = SYSCALL_BLOCK,
+                .cond = syscallcondition_new(trigger, need_timer ? sys->timer : NULL)};
+        }
+    }
+
+    // We have events now and we've already written them to fds_ptr
+    debug("poll returning %i ready events now", num_ready);
+done:
+    // Clear epoll for the next poll
+    epoll_reset(sys->epoll);
+    return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = num_ready};
+}
+
+static int _syscallhandler_checkPollArgs(PluginPtr fds_ptr, nfds_t nfds) {
+    if (nfds >= INT_MAX) {
+        debug("nfds was out of range [0, INT_MAX], returning EINVAL");
+        return -EINVAL;
```

Comment:
Should this be just `nfds > INT_MAX`?

Reply:
INT_MAX was a somewhat arbitrary choice of a "system limit" that shadow will impose. But I suppose `>` makes as much and maybe more sense than `>=` here.